src/analysis/monotone-analyzer.h

```diff
@@ -0,0 +1,75 @@
+#ifndef wasm_analysis_monotone_analyzer_h
+#define wasm_analysis_monotone_analyzer_h
+
+#include <iostream>
+#include <queue>
+#include <vector>
+
+#include "cfg.h"
+#include "lattice.h"
+#include "wasm-traversal.h"
+
+namespace wasm::analysis {
+
+template<size_t N> struct MonotoneCFGAnalyzer;
+
+// A node which contains all the lattice states for a given CFG node.
+template<size_t N>
+struct BlockState : public UnifiedExpressionVisitor<BlockState<N>> {
+  BlockState(const BasicBlock* underlyingBlock);
+
+  void addPredecessor(BlockState* pred);
+  void addSuccessor(BlockState* succ);
+
+  BitsetPowersetLattice<N>& getFirstState();
```

Comment:
Why is this hardcoded on a particular lattice?

Reply:
I intended this mainly as a proof of concept, so I hardcoded a lot of liveness analysis specific things. I plan to properly template and abstract this in a subsequent PR. However, if you believe that it is more appropriate to template it here, I can also implement that.